spmd/compiler/fusion.py

```diff
@@ -162,22 +165,82 @@ def _scan_graph_for_fusion_elements(
     return element_list
 
 
-def run_comm_fusion(gm: fx.GraphModule) -> bool:
-    """main entry into remapping graph for all_reduce fusion"""
+def _remove_gradient_tensor_clones(
+    gm: fx.GraphModule, comm_type: CommType = CommType.allreduce
+) -> int:
+    """
+    Optimizes away any duplicate gradient tensor nodes from DTensor
+    comm insertion in the provided graph.
+
+    Returns - total count of clone tensor nodes removed
+    """
+
+    count_clones_removed = 0
+    sequence: list[Union[str, CommType]] = [
+        "clone",
+        "_tensor_constant0",
+        "_tensor_constant1",
+        comm_type,
+    ]
+
+    len_sequence: int = len(sequence) - 1
+    index: int = 0
+    clone_node: fx.Node = None  # type: ignore
+    comm_node: fx.Node = None  # type: ignore
+
+    for node in gm.graph.nodes:
+
+        if node.op == OP.PLACEHOLDER:
+            index = 0
+            continue
+
+        pattern = sequence[index]
+
+        if (
+            index == 0
+            and node.op == OP.CALL_FUNCTION
+            and node.name.startswith(pattern)
+        ):
+            clone_node = node
+            index += 1
+            continue
+
+        elif index < len_sequence and node.name.startswith(pattern):
+            index += 1
+            continue
+
+        elif (
+            index == len_sequence
+            and node.op == OP.CALL_FUNCTION
+            and node.name.startswith(pattern)
+        ):
+            # found matching clone/comm sequence...
+            grad_tensor_node = clone_node.args[0]
+            comm_node = node
+
+            comm_node.update_arg(0, [grad_tensor_node])
```

Comment:
I'd vote for the more generic solution as in #631, as this fusion seems more closely bound with aotautograd internal behavior, e.g., how it handles reduce ops and process group args. 

Let me add more comments on the other PR.